daemon/algod/api/server/v2/test/handlers_test.go

```diff
@@ -693,21 +696,374 @@ func TestPostTransaction(t *testing.T) {
+						Format: nil, // should default to JSON
+					},
+					handle: protocol.JSONStrictHandle,
+				},
+			}
+
+			for _, responseFormat := range responseFormats {
+				t.Run(string(responseFormat.name), func(t *testing.T) { //nolint:paralleltest // Uses shared testing env
+					body = bytes.NewReader(bodyBytes)
+					req := httptest.NewRequest(http.MethodPost, "/", body)
+					rec := httptest.NewRecorder()
+
+					e := echo.New()
+					c := e.NewContext(req, rec)
+
+					// simulate transaction
+					err := handler.SimulateTransaction(c, responseFormat.params)
+					require.NoError(t, err)
+					require.Equal(t, 200, rec.Code, rec.Body.String())
+
+					// decode actual response
+					var actualBody model.SimulationResponse
+					decoder := codec.NewDecoderBytes(rec.Body.Bytes(), responseFormat.handle)
+					err = decoder.Decode(&actualBody)
+					require.NoError(t, err)
+
+					var expectedFailedAt *[]uint64
+					if len(scenario.FailedAt) != 0 {
+						clone := make([]uint64, len(scenario.FailedAt))
+						copy(clone, scenario.FailedAt)
+						clone[0]++
+						expectedFailedAt = &clone
+					}
+					expectedBody := model.SimulationResponse{
+						Version: 1,
+						TxnGroups: []model.SimulationTransactionGroupResult{
+							{
+								FailedAt: expectedFailedAt,
+								TxnResults: []model.SimulationTransactionResult{
+									{
+										TxnResult: makePendingTxnResponse(t, transactions.SignedTxnWithAD{
+											SignedTxn: stxns[0],
+											// expect no ApplyData info
+										}, responseFormat.handle),
+									},
+									{
+										TxnResult: makePendingTxnResponse(t, transactions.SignedTxnWithAD{
+											SignedTxn: stxns[1],
+											ApplyData: scenario.ExpectedSimulationAD,
+										}, responseFormat.handle),
+									},
+								},
+							},
+						},
+						WouldSucceed: scenario.Outcome == mocktracer.ApprovalOutcome,
+					}
+					assertSimulationResultsEqual(t, scenario.ExpectedError, expectedBody, actualBody)
+				})
+			}
+		})
+	}
+}
+
+func TestSimulateTransactionVerificationFailure(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// prepare node and handler
+	numAccounts := 5
+	offlineAccounts := true
+	mockLedger, roots, _, _, releasefunc := testingenv(t, numAccounts, 1, offlineAccounts)
+	defer releasefunc()
+	dummyShutdownChan := make(chan struct{})
+	mockNode := makeMockNode(mockLedger, t.Name(), nil, false)
+	mockNode.config.EnableExperimentalAPI = true
+	handler := v2.Handlers{
+		Node:     mockNode,
+		Log:      logging.Base(),
+		Shutdown: dummyShutdownChan,
+	}
+
+	hdr, err := mockLedger.BlockHdr(mockLedger.Latest())
+	require.NoError(t, err)
+	txnInfo := simulationtesting.TxnInfo{LatestHeader: hdr}
+
+	sender := roots[0]
+	receiver := roots[1]
+
+	txn := txnInfo.NewTxn(txntest.Txn{
+		Type:     protocol.PaymentTx,
+		Sender:   sender.Address(),
+		Receiver: receiver.Address(),
+		Amount:   0,
+	})
+
+	stxn := txn.Txn().Sign(sender.Secrets())
+	// make signature invalid
```

Comment:
I looked at the `simulation_eval_test.go` and saw a similar check - just to confirm, we error for invalid signatures first, but only report for missing signatures? I'm wondering why we make a distinction between the two.

Reply:
That's correct. Since we expect the simulation endpoint will be used for testing/debugging, we want to allow a fast path to enable quick iteration, which means either allowing missing or incorrect signatures.

We thought that allowing missing signatures was a more deliberate way for a developer to indicate that they don't care about having a valid signature. And it turns out this was much easier to implement as well, since it doesn't require modifying our core validation logic.